src/net.cpp

```diff
@@ -1994,11 +1999,30 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
             if (nTries > 100)
                 break;
 
-            CAddrInfo addr = addrman.SelectTriedCollision();
+            CAddrInfo addr;
 
-            // SelectTriedCollision returns an invalid address if it is empty.
-            if (!fFeeler || !addr.IsValid()) {
-                addr = addrman.Select(fFeeler);
+            if (fFeeler) {
+                // First, try to get a tried table collision address. This returns
+                // an empty (invalid) address if there are no collisions to try.
+                addr = addrman.SelectTriedCollision();
+
+                if (!addr.IsValid()) {
+                    // No tried table collisions. Select a new table address
+                    // for our feeler.
+                    addr = addrman.Select(true);
+                } else if (AlreadyConnectedToAddress(addr)) {
+                    // If test-before-evict logic would have us connect to a
+                    // peer that we're already connected to, just mark that
+                    // address as Good(). We won't be able to initiate the
+                    // connection anyway, so this avoids inadvertently evicting
+                    // a currently-connected peer.
+                    addrman.Good(addr);
```

Comment:
This will call `Good()` for any address that we're connected to, including inbound peers. That's different from the current behavior, where we're careful to not call `Good()` for inbounds. The result of that is we'll move addresses of inbound peers to the tried table, even though we've never actually successfully connected out to them.

Does this open up a gap where an attacker can get addresses added to your tried table by opening lots of inbound connections to try to trigger this logic?

Reply:
That can only happen if there's a tried-table collision with the address of the inbound peer -- so that inbound peer's address is already in our tried-table anyway...  I don't think calling `Good()` on an address already in our tried-table can be all that detrimental?  I think there are two cases to consider:
- If we were to initiate an outbound connection to the address, it would accept it, and we would therefore mark it `Good()`.  In this case, marking it `Good()` because it's a current peer is correct.
- If we initiated an outbound connection to the address, it would not accept it.  In this case, it's sort of unclear what is going on; we must have been able to connect to it as an outbound in the past in order to get it into the tried table in the first place.  So perhaps the peer isn't accepting inbound connections at this time (or is full), or there's something funny going on with the networking/address that we're trying to use to reach it?  However, keeping it in our tried table doesn't seem detrimental; at most, about 115 addresses in our tried table could be reserved for us by inbound peers who no longer are taking incoming connections but were in the past...  An attacker could just keep listening on those addresses anyway and maintain those tried-table entries (tried-table collisions are rare anyway), so I don't know that we're giving an attacker much more power or reducing the costs of an eclipse attack all that much by failing to detect this case.

Another option would be to (a) change our outbound peer logic to allow feeler connections to inbound peers, and (b) change this logic to mark outbound peers' addresses as `Good()` in the event of a tried table collision.  I am not sure if there's some reason that would be a bad idea, but I'm also not sure that it's necessary.

Reply:
Ah yes, of course you're right. Anything returned from `SelectTriedCollision()` must by definition already be in the tried table.